polyval/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+    buffer: [u8; 16],
+    o: PhantomData<B>,
+}
+
+const XMM_MASK: u128 = 0x0001_c200_0000_0000_0000;
+
+// TODO(tarcieri): POLYVAL is a universal hash function, not a MAC.
+// It can be used to implement a MAC (e.g. GMAC), but should probably use
+// a different trait than `crypto_mac::Mac`.
+impl<B> Mac for PolyvalCore<B>
+where
+    B: ByteOrder,
+{
+    type KeySize = U16;
+    type OutputSize = U16;
+
+    fn new(h: &GenericArray<u8, Self::KeySize>) -> Self {
+        Self {
+            h: B::read_u128(h),
+            s: 0,
+            leftover: 0,
+            buffer: [0u8; 16],
+            o: PhantomData,
+        }
+    }
+
+    #[inline]
+    fn input(&mut self, data: &[u8]) {
+        let mut m = data;
+
+        if self.leftover > 0 {
+            let want = min(16 - self.leftover, m.len());
+
+            for (i, byte) in m.iter().cloned().enumerate().take(want) {
+                self.buffer[self.leftover + i] = byte;
+            }
+
+            m = &m[want..];
+            self.leftover += want;
+
+            if self.leftover < 16 {
+                return;
+            }
+
+            let tmp = self.buffer;
+            self.block(&tmp);
+            self.leftover = 0;
+        }
+
+        while m.len() >= 16 {
+            self.block(&m[..16]);
+            m = &m[16..];
+        }
+
+        self.buffer[..m.len()].copy_from_slice(m);
+        self.leftover = m.len();
+    }
+
+    fn result(mut self) -> MacResult<Self::OutputSize> {
+        // TODO(tarcieri): this is almost certainly wrong
+        if self.leftover > 0 {
+            self.buffer[self.leftover] = 1;
+            for i in self.leftover + 1..16 {
+                self.buffer[i] = 0;
+            }
+
+            let tmp = self.buffer;
+            self.block(&tmp);
+        }
+
+        let mut mac = GenericArray::default();
+        B::write_u128(&mut mac, self.s);
+        MacResult::new(mac)
+    }
+
+    fn reset(&mut self) {
+        self.s = 0;
+        self.buffer = [0u8; 16];
+        self.leftover = 0;
+    }
+}
+
+impl<B> PolyvalCore<B>
+where
+    B: ByteOrder,
+{
+    fn block(&mut self, block: &[u8]) {
+        debug_assert_eq!(block.len(), 16);
+        let x = self.s ^ B::read_u128(block);
+
+        let t1 = pclmulqdq(x, self.h, 0x00);
+        let t2 = pclmulqdq(x, self.h, 0x01);
+        let t3 = pclmulqdq(x, self.h, 0x10);
+        let t4 = pclmulqdq(x, self.h, 0x11);
+        let t5 = t2 ^ t3;
+
+        let mut t6 = t1 ^ ((t5 & 0xffff_ffff) << 64);
+        t6 = pclmulqdq(XMM_MASK, t6, 0x01) ^ t6.rotate_left(64);
+        t6 = pclmulqdq(XMM_MASK, t6, 0x01) ^ t6.rotate_left(64);
+
+        self.s = t4 ^ (t5 >> 64) ^ t6
+    }
+}
+
+#[cfg(feature = "zeroize")]
+impl<O> Drop for PolyvalCore<O>
+where
+    O: ByteOrder,
+{
+    fn drop(&mut self) {
+        self.h.zeroize();
+        self.s.zeroize();
+    }
+}
+
+/// Software reimplementation of the PCLMULQDQ CPU instruction
+// TODO(tarcieri): support for CLMUL intrinsics
```

Comment:
Sure would be good to actually add these because otherwise performance seems abysmal